0035-feature-flags/README.md

```diff
@@ -0,0 +1,418 @@
+-->
+
+# ZEP-0035: Feature Flags
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+      - [Unit tests](#unit-tests)
+      - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+This ZEP proposes Feature Flags for Zarf, intended to provide a configuration model for features throughout the stages
+of release and deprecation, eg. "alpha", "beta", "GA", "deprecated". It introduces a ctx-based API, similar to previous
+usages of feature flags in Zarf, a centralized location where all CLI feature flags are **declared** so that users may
+easily find Flags and their documentation in code. Additionally, centralizing where Flags are declared allows us to
+automate site documentation that corresponds one to one with the docs in code. Inspiration for the Flag fields is taken
+from the Kubernetes project.
+
+## Motivation
+
+Feature flags have been discussed often in Zarf as a step on the path to v1.0.0 and we believe they provide significant
+enough benefits to offset the complexity they add. Maintainer flexibility is the foremost benefit, allowing for 
+experimental features to be added and iterated on. Allowing users to opt-in for alpha previews and beta testing also
+provides additional ways for Zarf maintainers to get feedback and shape features into a complete state before they are
+made fully available.
+
+There's also documentation benefits. By associating features to their release version, users may
+more easily track which minimum version is required to get the functionality they need. Currently, this information can
+be dug up in GitHub releases and commits, but is not modeled explicitly in the codebase.
+
+Lastly, Zarf's ZEP process is deeply tied to a staged (alpha, beta, GA) release flow, and feature flags offer an
+opportunity to model and encourage that flow in the codebase itself.
+
+### Goals
+
+- Make new features opt-in.
+- Have a clear and documented reference for users on which versions of Zarf introduce new features.
+- Runtime-level feature configuration
+- Stretch goal: multiple avenues of configuration (code, env vars, flags, config files)
+
+### Non-Goals
+
+- Elaborate configuration nightmares, where users don't know what is enabled or multiple flags are intertwined.
+- Forever features that are endless reworked and never make it to GA.
```

Comment:
I'm inclined to move this to goals, to ensure we don't have forever features. Ideally you want to enforce some kind of timeline for every feature, but if you see something is stuck for too long you might want to consider to drop it, which is also easier depending on the stage the features is at. Alpha features can be dropped almost at a whim, whereas beta (especially those enabled by default) or stable require a proper deprecation period to allow users to stop using those.